Add a JSON pretty print function

PhictionalOne opened the pull request:
relates to https://github.com/duckdb/duckdb/discussions/12388

`json_pretty` behaves similar to `json_serialize_plan` and only prettifies the `VARCHAR` and _not_ the cli output.

```
D .mode duckbox
D SELECT json_pretty('[1,2,   {"a": "Hi"}]');
┌───────────────────────────────────────────────────────┐
│          json_pretty('[1,2,   {"a": "Hi"}]')          │
│                        varchar                        │
├───────────────────────────────────────────────────────┤
│ [\n    1,\n    2,\n    {\n        "a": "Hi"\n    }\n] │
└───────────────────────────────────────────────────────┘
D .mode ascii
D SELECT json_pretty_print('[1,2,   {"a": "Hi"}]');
json_pretty_print('[1,2,   {"a": "Hi"}]')[
    1,
    2,
    {
        "a": "Hi"
    }
]D 
```

PhictionalOne committed:
Add json_pretty_print function

PhictionalOne committed:
Add unit test

PhictionalOne committed:
Add to catalog

Tishj commented on extension/json/json_functions/json_pretty_print.cpp:
```diff
@@ -0,0 +1,32 @@
+	idx_t len;
+	auto data =
+	    yyjson_val_write_opts(val, JSONCommon::WRITE_PRETTY_FLAG, alc, reinterpret_cast<size_t *>(&len), nullptr);
+	return string_t(data, len);
```
I'm not seeing StringVector::AddString anywhere, which worries me
`string_t` does not take ownership of the memory.

Which is why we often use StringVector::AddString, which copies it to the StringHeap that's owned by the Vector, guaranteeing that the string_t we return will point to memory that lives as long as the Vector that's holding it

Tishj replied:
This is also used in `json_serialize_plan` in the file that I referenced

PhictionalOne committed:
Fix catalog order

Mytherin commented on test/sql/json/scalar/test_json_pretty_print.test:
```diff
@@ -0,0 +1,39 @@
+statement ok
+insert into test values ('{}'), ('[]'), ('[1, 2, 3]')
+
+mode output_hash
```
Was this left in by accident?

PhictionalOne replied:
Yes, I removed the test table now.

Mytherin replied:
I meant the `mode output_hash`, not the test table. Adding `mode output_hash` prevents the test from actually testing anything

Mytherin reviewed (commented):
Thanks! LGTM - perhaps @lnkuiper also wants to have a look

PhictionalOne committed:
Remove irrelevant table from test

Mytherin commented:
In Postgres the equivalent function is called [`jsonb_pretty`](https://www.postgresql.org/docs/9.5/functions-json.html), and in SQLite it is called `[json_pretty`](https://www.sqlite.org/json1.html#jpretty). Could we rename this function to `json_pretty` for consistency with the other systems?

PhictionalOne committed:
Rename to json_pretty

Tishj commented on extension/json/json_functions/json_pretty.cpp:
```diff
@@ -0,0 +1,32 @@
+	auto json_type = args.data[0].GetType();
+	D_ASSERT(json_type == LogicalType::VARCHAR || json_type == LogicalType::JSON());
+
+	JSONExecutors::UnaryExecute<string_t>(args, state, result, PrettyPrint);
```
Should we pretty-print nulls?
I believe the UnaryExecutor behaves as null in -> null out, not calling the function.

I guess nulls are represented differently in json, so maybe this is not a problem?

PhictionalOne replied:
`NULL`s in JSON are different from SQL nulls.

As far as I can see the following behaviour should be the expected one:
```sql
D .mode duckbox
D SELECT json_pretty(null);
┌───────────────────┐
│ json_pretty(NULL) │
│      varchar      │
├───────────────────┤
│                   │
└───────────────────┘
D SELECT json_pretty('null'::json);
┌─────────────────────────────────────┐
│ json_pretty(CAST('null' AS "json")) │
│               varchar               │
├─────────────────────────────────────┤
│ null                                │
└─────────────────────────────────────┘
D .mode ascii
D SELECT json_pretty('[1,2,null, {"a": null},  42]'::json);
json_pretty(CAST('[1,2,null, {"a": null},  42]' AS "json"))[
    1,
    2,
    null,
    {
        "a": null
    },
    42
]D 
```

I would include them all in the unit test.

Tishj commented on test/sql/json/scalar/test_json_pretty.test:
```diff
@@ -0,0 +1,33 @@
+# name: test/sql/json/scalar/test_json_pretty.test
+# description: Test JSON pretty printer
```
Can we add a test with `NULL` and also nulls inside nested json objects?

lnkuiper commented on extension/json/json_functions/json_pretty.cpp:
```diff
@@ -0,0 +1,32 @@
+	idx_t len;
+	auto data =
+	    yyjson_val_write_opts(val, JSONCommon::WRITE_PRETTY_FLAG, alc, reinterpret_cast<size_t *>(&len), nullptr);
+	return StringVector::AddString(result, data, len);
```
You can simply return `string_t(data, len)` here, as the string will stay cached for long enough in the `alc` until the next vector is processed. Using `StringVector::AddString` here will unnecessarily create an extra copy of the string.

PhictionalOne replied:
Ah. I had that before (https://github.com/duckdb/duckdb/pull/12398#discussion_r1627980229). So I should roll back that change?

lnkuiper replied:
Yes, it can be rolled back. We don't have to add the string to the vector since it will be cached in the allocator. We're doing the exact same thing in `json_extract.cpp` and `json_transform.cpp`. Sorry for the confusion.

lnkuiper reviewed (changes requested):
Thanks for the contribution! I have just one comment, then it's good to go from my side:

PhictionalOne commented on extension/json/json_functions/json_pretty.cpp:
```diff
@@ -0,0 +1,32 @@
+	D_ASSERT(alc);
+	idx_t len;
+	auto data =
+	    yyjson_val_write_opts(val, JSONCommon::WRITE_PRETTY_FLAG, alc, reinterpret_cast<size_t *>(&len), nullptr);
```
Honestly, I didn't thought of that as it is copy-pasted from `json_serialize_sql`. Is there a resource where I could look that up?

PhictionalOne committed:
Rewrite unit test

- Move from hash to string comparison
- Introduce three test cases using SQL and JSON NULLs

Mytherin commented:
Thanks!
